pkg/controller/nodelifecycle/node_lifecycle_controller.go

```diff
@@ -787,19 +794,29 @@ func (nc *Controller) monitorNodeHealth() error {
 				nc.processNoTaintBaseEviction(node, &observedReadyCondition, gracePeriod)
 			}
 
-			// Report node event.
-			if currentReadyCondition.Status != v1.ConditionTrue && observedReadyCondition.Status == v1.ConditionTrue {
-				nodeutil.RecordNodeStatusChange(nc.recorder, node, "NodeNotReady")
+			_, needsRetry := nc.nodesToRetry.Load(node.Name)
+			// We call markPodsNotReadyFunc when node condition changed from true to something different
+			// or previous markPodsNotReadyFunc failed and ready status is still not true.
+			if (currentReadyCondition.Status != v1.ConditionTrue && observedReadyCondition.Status == v1.ConditionTrue) ||
+				(needsRetry && observedReadyCondition.Status != v1.ConditionTrue) {
+				// Report node event only once when status changed.
+				if !needsRetry {
```

Comment:
Those conditions became fairly complicated...

How about factoring this out to a separate function:
```
func (...) markPodsNotReady() {
  pods, err := listPodsFromNode(nc.kubeClient, node.Name)
  ...
```

And here doing:
```
if currentReadyCondition.. && observed... {
  // report node event
  nc.markPodsNotReady()
} else if needsRetry && observed... {
  nc.markPodsNodReady()
}
```

Reply:
I don't want to log error or handle nodesToRetry in markPodsNodReady but rather in parent method.
I went with swtich/case. Is `fallthrough` acceptable or we try to not use in the code?